### PR TITLE
[ LAYER ] Add backwarding of Embedding Layer

### PR DESCRIPTION
**Pull request by jijoongmoon**
This PR includes
 - Backwarding implementation of Embedding Layer
 - Add Test Cases ( More Test Cases will be added )

**Self evaluation:**
1. Build test:	 [X]Passed [ ]Failed [ ]Skipped
2. Run test:	 [X]Passed [ ]Failed [ ]Skipped

Signed-off-by: jijoong.moon <jijoong.moon@samsung.com>

### CODE REVIEW

File: nntrainer/layers/embedding.cpp [hunk 1]
```diff
@@ -104,17 +104,19 @@ void EmbeddingLayer::forwarding(bool training) {
     float *in_data = input_.getAddress(b * input_.getDim().getFeatureLen());
 
     for (unsigned int i = 0; i < in_length; ++i) {
-      if (in_data[i] < 0)
+      if (in_data[i] > in_dim) {
```
Comment (zhoonit) [inline nntrainer/layers/embedding.cpp:107]: if (`in_data[i] == in_dim`) is allowed, I wonder if `in_data[i] == 0` is defined in this case? for keras, only positive integer seems allowed.

File: nntrainer/layers/embedding.cpp [hunk 2]
```diff
@@ -104,17 +104,19 @@ void EmbeddingLayer::forwarding(bool training) {
...
+        throw std::invalid_argument("input word index is greater then in_dim");
```
Comment (zhoonit) [inline nntrainer/layers/embedding.cpp:108]:
```suggestion
        throw std::invalid_argument("input word index is greater than in_dim");
```

File: nntrainer/layers/embedding.cpp [hunk 3]
```diff
@@ -104,17 +104,19 @@ void EmbeddingLayer::forwarding(bool training) {
...
+      }
+
+      if (in_data[i] - 1 < 0)
```
Comment (zhoonit) [inline nntrainer/layers/embedding.cpp:111]: if this is not a hard error, I guess we have to fill the output to zero?

File: test/unittest/unittest_nntrainer_layers.cpp [hunk 4]
```diff
@@ -2260,6 +2260,46 @@ TEST_F(nntrainer_EmbeddingLayer, forwarding_01_p) {
...
+  for (unsigned int i = 0; i < derivatives.getDim().getDataLen(); ++i) {
+    derivatives.getData()[i] = 1.0;
+  }
+
+  setOptimizer(nntrainer::OptType::ADAM, "learning_rate=1.0");
+
+  allocateMemory();
+
+  nntrainer::Tensor result;
+  EXPECT_NO_THROW(result = *layer.backwarding_with_val(
```
Comment (zhoonit) [inline test/unittest/unittest_nntrainer_layers.cpp:2299]: Is the test matching  numbers? if not could you please left a comment?

**Review by zhoonit — APPROVED** [review]
LGTM

File: nntrainer/layers/embedding.cpp (hunk 3 above)
Comment (kparichay) [reply nntrainer/layers/embedding.cpp:111]: Does the indexing of data start from 0 or 1? If 0, then `in_data[i] == 0` is valid.

Comment (kparichay) [inline nntrainer/layers/embedding.cpp:111]: Does the indexing of data start from 0 or 1? If 0, then `in_data[i] == 0` is valid.

File: nntrainer/layers/embedding.cpp [hunk 5]
```diff
@@ -133,10 +135,38 @@ void EmbeddingLayer::copy(std::shared_ptr<Layer> l) {
 }
 
 void EmbeddingLayer::calcDerivative() {
-  // NYI
+  return; // intended
```
Comment (kparichay) [inline nntrainer/layers/embedding.cpp:138]: It seems like embedding layer does not pass the derivative back. If yes, then please throw here.

File: nntrainer/layers/embedding.cpp [hunk 6]
```diff
@@ -133,10 +135,38 @@ void EmbeddingLayer::copy(std::shared_ptr<Layer> l) {
...
+
+  // TODO:
+  // This is to calculate gradient with current implementation of optimizer.
+  // In order to accelerate, we need to better way like using index to weight.
+
+  for (unsigned int b = 0; b < input_.batch(); ++b) {
+    float *in_data = input_.getAddress(b * input_.getDim().getFeatureLen());
+
+    for (unsigned int i = 0; i < in_length; ++i) {
+      if (in_data[i] - 1 < 0)
```
Comment (kparichay) [inline nntrainer/layers/embedding.cpp:157]: This is not needed. This check was already verified in `forwarding`.

File: nntrainer/layers/embedding.cpp [hunk 7]
```diff
@@ -133,10 +135,38 @@ void EmbeddingLayer::copy(std::shared_ptr<Layer> l) {
...
+        continue;
+
+      float *djdw_data =
+        djdw.getAddress(static_cast<uint>(in_data[i] - 1) * out_dim);
+      float *grad_data = derivative_.getAddress(
+        b * derivative_.getDim().getFeatureLen() + i * out_dim);
+
+      for (unsigned int j = 0; j < out_dim; ++j) {
```
Comment (kparichay) [inline nntrainer/layers/embedding.cpp:165]: please use memset here instead of a loop.

File: nntrainer/layers/embedding.cpp [hunk 8]
```diff
@@ -133,10 +135,38 @@ void EmbeddingLayer::copy(std::shared_ptr<Layer> l) {
...
+  Tensor &djdw =
+    weightAt(static_cast<int>(EmbeddingParams::weight)).getGradientRef();
+  Tensor &derivative_ = net_hidden[0]->getGradientRef();
+  Tensor &input_ = net_input[0]->getVariableRef();
+
+  djdw.setZero();
+
+  // TODO:
+  // This is to calculate gradient with current implementation of optimizer.
+  // In order to accelerate, we need to better way like using index to weight.
```
Comment (kparichay) [inline nntrainer/layers/embedding.cpp:151]: :100:

File: nntrainer/layers/embedding.cpp (hunk 2 above)
Comment (jijoongmoon) [reply nntrainer/layers/embedding.cpp:108]: done.

File: nntrainer/layers/embedding.cpp (hunk 3 above)
Comment (jijoongmoon) [reply nntrainer/layers/embedding.cpp:111]: It must be skipped.  It is just padding. No need to calculate.

Comment (jijoongmoon) [reply nntrainer/layers/embedding.cpp:111]: It is padding and padding could be less or equal zero.

File: nntrainer/layers/embedding.cpp (hunk 5 above)
Comment (jijoongmoon) [reply nntrainer/layers/embedding.cpp:138]: No. It need to be update the gradient. so.. It is not like input layer. Input Layer does not need to do backwarding. But we do not have detail control about it. This need another change but.. for not it is ok as it is. ( You can see line 353 in neuralnet.cpp )

File: nntrainer/layers/embedding.cpp (hunk 6 above)
Comment (jijoongmoon) [reply nntrainer/layers/embedding.cpp:157]: It is needed. it is for the skipping of calculation.

File: nntrainer/layers/embedding.cpp (hunk 7 above)
Comment (jijoongmoon) [reply nntrainer/layers/embedding.cpp:165]: It is not assign. It addition.

File: test/unittest/unittest_nntrainer_layers.cpp (hunk 4 above)
Comment (jijoongmoon) [reply test/unittest/unittest_nntrainer_layers.cpp:2291]: I will add another PR for test number after PR#1002 and #1004

File: nntrainer/layers/embedding.cpp (hunk 3 above)
Comment (kparichay) [reply nntrainer/layers/embedding.cpp:111]: `in_data` is the input data. I wonder how is it padding? can you please elaborate on the padding part?
I also wonder why the embedding layer needs padding.

File: nntrainer/layers/embedding.cpp (hunk 5 above)
Comment (kparichay) [reply nntrainer/layers/embedding.cpp:138]: This layer calculates the derivative for the next layer in backpropagation.
In order for the network to run correctly, either this layer must give back the derivative or throw to ensure that the layer behind it (if any) does not receive any garbage gradient.
Just a simple return invalidates the invariant of the network training.

File: nntrainer/layers/embedding.cpp (hunk 6 above)
Comment (kparichay) [reply nntrainer/layers/embedding.cpp:157]: Can you please add in the comments the scenario where in_data[i] is less than 1?
Also, can you write the semantics of in_data - that it must be an unsigned integer etc.

File: nntrainer/layers/embedding.cpp (hunk 7 above)
Comment (kparichay) [reply nntrainer/layers/embedding.cpp:165]: then please use something like `std::transform (djdw.begin(), djdw.end(), grad_data.begin(), djdw.begin(), std::plus<float>());`

File: nntrainer/layers/embedding.cpp (hunk 3 above)
Comment (jijoongmoon) [reply nntrainer/layers/embedding.cpp:111]: well.. the words is indexed as number. and the input is sentence. the maximum size of words is in_length.. and sometimes it is shorter than this. when it is short.. we need to fill out with padding. 
' Well done!' could be [ 1, 2, 0, 0] with in_length is 4.

File: nntrainer/layers/embedding.cpp (hunk 5 above)
Comment (jijoongmoon) [reply nntrainer/layers/embedding.cpp:138]: I will leave another issue for this. but not this pr.

File: nntrainer/layers/embedding.cpp (hunk 7 above)
Comment (jijoongmoon) [reply nntrainer/layers/embedding.cpp:165]: does it faster than this? if it is, let me know.

File: nntrainer/layers/embedding.cpp (hunk 6 above)
Comment (jijoongmoon) [reply nntrainer/layers/embedding.cpp:157]: Like I describe earlier, it is padding.

File: nntrainer/layers/embedding.cpp (hunk 3 above)
Comment (kparichay) [reply nntrainer/layers/embedding.cpp:111]: Then how making this check `in_data[i] == 0`?

File: nntrainer/layers/embedding.cpp (hunk 7 above)
Comment (kparichay) [reply nntrainer/layers/embedding.cpp:165]: yes, it is faster with `std::transform()` as it is easier for compile to optimize these, and from c++17, allows parallel computation with transform.

File: nntrainer/layers/embedding.cpp (hunk 5 above)
Comment (kparichay) [reply nntrainer/layers/embedding.cpp:138]: Please raise a issue for this :)

File: nntrainer/layers/embedding.cpp (hunk 3 above)
Comment (jijoongmoon) [reply nntrainer/layers/embedding.cpp:111]: It also can be -1.

Comment (kparichay) [reply nntrainer/layers/embedding.cpp:111]: Can you please add the possible values and what they denote as comments? BTW, what does -1 denote?

File: nntrainer/layers/embedding.cpp (hunk 7 above)
Comment (jijoongmoon) [reply nntrainer/layers/embedding.cpp]: done

File: nntrainer/layers/embedding.cpp (hunk 3 above)
Comment (jijoongmoon) [reply nntrainer/layers/embedding.cpp:111]: I guess It depends how user set. As you know, index starts 0 or 1. right? If index starts 0, then it will -1. if it starts from 1, the padding will be 0. Nothing more.

File: nntrainer/layers/embedding.cpp [hunk 9]
```diff
@@ -133,10 +135,41 @@ void EmbeddingLayer::copy(std::shared_ptr<Layer> l) {
...
+
+  for (unsigned int b = 0; b < input_.batch(); ++b) {
+    float *in_data = input_.getAddress(b * input_.getDim().getFeatureLen());
+
+    for (unsigned int i = 0; i < in_length; ++i) {
+      if (in_data[i] - 1 < 0)
+        continue;
```
Comment (lhs8928) [inline nntrainer/layers/embedding.cpp:162]: Cause if padding is exist then it will be 0 until the end of in_data, so I think break will be better than continue.

File: nntrainer/layers/embedding.cpp (hunk 3 above)
Comment (kparichay) [reply nntrainer/layers/embedding.cpp:111]: Then this condition will come out wrong if the actual index starts from 0, as this condition discards them too.

**Review by lhs8928 — APPROVED** [review]
LGTM

File: nntrainer/layers/embedding.cpp (hunk 9 above)
Comment (kparichay) [reply nntrainer/layers/embedding.cpp:166]: when using break, you also need to think about masking the inputs for future.

**Review by kparichay — APPROVED** [review]
LGTM
